### ISSUE TITLE
Windows USB support through Qemu

### ISSUE THREAD

**Issue by lucaszanella** [QubesOS/qubes-issues#5802]
Since Windows always runs in HVM mode, which uses Qemu, shouldn't it be possible to emulate USB devices on Windows through Qemu?

As I understand, for linux, paravirtualization drivers are used to send and receive USB packets through the network. I guess the same could be done on Windows, couldn't it? Sending USB packets through network to Qemu and emulate it to appear in Windows. Could even be useful for having USB on any OS in HVM mode, even very exotic ones.

I really like linux but sometimes I have to work with Windows, so it'd be very great to have a totally functional Windows 10 inside Qubes

**Comment by marmarek** [comment]
Yes, that could work. See also https://www.qubes-os.org/gsoc/#usb-passthrough-to-windows-qubes

It's "just" that someone needs to put it all together.

**Comment by lucaszanella** [comment]
Thanks, I didn't know this page. Lots of great ideas. I'm studying Qemu internals. Maybe one day I can help :)

Enviado via ProtonMail móvel

-------- Mensagem Original --------
Ativo 7 de mai de 2020 21:52, Marek Marczykowski-Górecki escreveu:

**Comment by brendanhoar** [comment]
An alternative to emulated USB via QEMU is to utilize the capability to attach the PCI device that is the PCI USB controller directly to HVMs, including Windows ones (see qvm-pci). This is mostly useful if your host machine has, or can have, multiple USB controllers connected to it.

E.g.

On the laptop I use, I have two different internal USB controllers (one for USB 3.0 and one for USB 2.0) on different PCI devices. In addition, I have an expresscard USB 3.0 controller installed (and the expresscard port enabled in BIOS). 

I typically connect the USB 2.0 controller to sys-usb and leave the USB 3.0 controllers unassigned so that I can connect the USB 3.0 controllers to HVMs as needed.

B

**Comment by easydozen** [comment]
I undestand that there are a number of priority issues and don't want to push on qubes team here. But actually, I've already have more or less working PoC solution. If someone wants to check it out take a look at winusb branch in my personal repos:

https://github.com/easydozen/qubes-vmm-xen-stubdom-linux/tree/winusb
https://github.com/easydozen/qubes-app-linux-usb-proxy/tree/winusb
https://github.com/easydozen/qubes-core-qrexec/tree/winusb
https://github.com/easydozen/qubes-core-admin/tree/winusb

Additionally, xen package should be rebuilt with increased max_maptrack_frames value:
`
sed -i -e s/"max_maptrack_frames = 0"/"max_maptrack_frames = guest_config->b_info.max_maptrack_frames"/g tools/libxl/libxl_dm.c`

**Comment by marmarek** [comment]
> Additionally, xen package should be rebuilt with increased max_maptrack_frames value:
> ` sed -i -e s/"max_maptrack_frames = 0"/"max_maptrack_frames = guest_config->b_info.max_maptrack_frames"/g tools/libxl/libxl_dm.c`

Can you send a patch to xen-devel?

**Comment by easydozen** [comment]
> > Additionally, xen package should be rebuilt with increased max_maptrack_frames value:
> > ` sed -i -e s/"max_maptrack_frames = 0"/"max_maptrack_frames = guest_config->b_info.max_maptrack_frames"/g tools/libxl/libxl_dm.c`
> 
> Can you send a patch to xen-devel?

https://lists.xenproject.org/archives/html/xen-devel/2020-09/msg00949.html

**Comment by easydozen** [comment]
The suggested implementation supports wide spectre of usb devices using _nec-qemu-xhci_ controller. It's thoroughly tested and seems stable.

There are several qubes projects involved in the implementation: _xen_, _stubdom_, _libvirt_, _core-admin_, _core-qrexec_, _usb-proxy_.

Unfortunately aforementioned patch to _xen_ wasn't merged in _xen-devel_, so for now I've made it here as a PR.

As discussed earlier at IRC, I prepared two variants of _stubdom_ - minimal and full. There is usbip, libusb (w/o udev), pulseaudio, qrexec-agent (w/o pam) support in the second case. Thus, changes in the _stubdom_ project among other things affects an audio support (QubesOS/qubes-issues#2624). User can switch to the full variant using alternative emulator' ramdisk and kernel. The corresponding patch to _libvirt_ makes it affordable. PRs to _core-admin_ and _usb-proxy_ allows rpc calls redirection to _qrexec-agent_ in the stubdom.

**Comment by jandryuk** [comment]
@easydozen Nice work!  Have you tried passing through a webcam, and does it perform acceptably?

**Comment by easydozen** [comment]
> @easydozen Nice work! Have you tried passing through a webcam, and does it perform acceptably?

Qemu docs says that there is no support for real time streaming devices in usb-host emulation, but actually it depends on your hardware, qemu controller and guest drivers. I tried several integrated webcams some of them work flawlessly, but I saw glitches and freezes too. 
To avoid problems with guest drivers in windows systems nec-usb-xhci controller was chosen. This is the only option for win7. 
To make the specific usb device work in win10 switching to qemu-xhci or ehci controller may help.

Win10 provides support for unix sockets, so for this system, maybe, it would be better to implement the same approach as in linux guest.
